service/storage_service.cc

```diff
@@ -831,6 +831,21 @@ storage_service::get_range_to_address_map(const sstring& keyspace,
     return construct_range_to_endpoint_map(ks, get_all_ranges(sorted_tokens));
 }
 
+void storage_service::handle_state_replacing_update_pending_ranges(inet_address replacing_node) {
+    try {
+        slogger.info("handle_state_replacing: Waiting for replacing node {} to be alive on all shards", replacing_node);
+        _gossiper.wait_alive({replacing_node}, std::chrono::milliseconds(5 * 1000));
+        slogger.info("handle_state_replacing: Replacing node {} is now alive on all shards", replacing_node);
+    } catch (...) {
+        slogger.warn("handle_state_replacing: Failed to wait for replacing node {} to be alive on all shards: {}",
+                replacing_node, std::current_exception());
+    }
+    auto tmptr = get_mutable_token_metadata_ptr().get0();
```

Comment:
Should this be executed under `mutate_token_metadata()` to prevent corruption of token metadata on concurrent modification?

@bhalevy you're more familiar with the new metadata manipulation infra, please check.

Reply:
on_alive() is asynchronous to the gossiper.

Reply:
We should get the tmlock as we do in storage_service::handle_state_replacing.  I missed it when copying the "auto tmptr = get_mutable_token_metadata_ptr().get0()".

Reply:
> We should get the tmlock as we do in storage_service::handle_state_replacing. I missed it when copying the "auto tmptr = get_mutable_token_metadata_ptr().get0()".

yes, that's valid.  see `storage_service::handle_state_replacing` for example.
Bute `storage_service::mutate_token_metadata` is more elegant as it automates also `replicate_to_all_cores`.
So you could implement this block more simply like this:
```
mutate_token_metadata([this, replacing_node] (mutable_token_metadata_ptr tmptr) mutable {
    slogger.info("handle_state_replacing: Update pending ranges for replacing node {}", replacing_node);
    return update_pending_ranges(tmptr, format("handle_state_replacing {}", replacing_node));
}).get();
```

Reply:
I checked 4.0 does not have the mutate_token_metadata stuff. I will stick to replicate_to_all_cores to make backport easier.

Reply:
Since update_pending_ranges() is non-atomic, should we still have some sort of protection against concurrent updates from listeners?

The old listeners were fine because they were all run from the gossiper thread, so serialized.

Reply:
I can see that you take the lock